pkg/declarative/v2/reconciler.go

```diff
@@ -0,0 +1,405 @@
+package v2
+
+import (
+	"context"
+	"errors"
+
+	manifestClient "github.com/kyma-project/module-manager/pkg/client"
+	manifestLabels "github.com/kyma-project/module-manager/pkg/labels"
+	"github.com/kyma-project/module-manager/pkg/types"
+	"github.com/kyma-project/module-manager/pkg/util"
+	"helm.sh/helm/v3/pkg/kube"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+func NewFromManager(mgr manager.Manager, prototype Object, options ...Option) reconcile.Reconciler {
+	r := &Reconciler{}
+	r.prototype = prototype
+	r.Options = DefaultOptions().Apply(WithManager(mgr)).Apply(options...)
+	return r
+}
+
+type Reconciler struct {
+	prototype Object
+	*Options
+}
+
+type ConditionType string
+
+const (
+	ConditionTypeResources    ConditionType = "Resources"
+	ConditionTypeInstallation ConditionType = "Installation"
+)
+
+type ConditionReason string
+
+const (
+	ConditionReasonResourcesAreAvailable ConditionReason = "ResourcesAvailable"
+	ConditionReasonReady                 ConditionReason = "Ready"
+)
+
+func newInstallationCondition(obj Object) metav1.Condition {
+	return metav1.Condition{
+		Type:               string(ConditionTypeInstallation),
+		Reason:             string(ConditionReasonReady),
+		Status:             metav1.ConditionFalse,
+		Message:            "installation is ready and resources can be used",
+		ObservedGeneration: obj.GetGeneration(),
+	}
+}
+
+func newResourcesCondition(obj Object) metav1.Condition {
+	return metav1.Condition{
+		Type:               string(ConditionTypeResources),
+		Reason:             string(ConditionReasonResourcesAreAvailable),
+		Status:             metav1.ConditionFalse,
+		Message:            "resources are parsed and ready for use",
+		ObservedGeneration: obj.GetGeneration(),
+	}
+}
+
+//nolint:funlen,nestif,gocognit,gocyclo,cyclop,maintidx //TODO discuss if worth breaking up or if more readable as is.
```

Comment:
For me: it would provide better readability if we split this function into sub-functions. It will also help in debugging.